File: crates/netcfgd-sys/src/lock.rs
//! An exclusive advisory lock, held for as long as the guard is alive.
//!
//! Here rather than beside the code that wants it, for the reason [`crate::term`]
//! gives about echo: constraint 4 confines `unsafe` to one audited crate, and
//! this is it -- not because it is the netlink crate, but because it is where
//! the libc boundary lives.
//!
//! **Why not `std::fs::File::lock`.** It exists and does exactly this, and it
//! is unstable: `file_lock` is not in 1.85, which is this workspace's
//! `rust-version` and also the rustc this machine has. Taking it would raise
//! the floor for everybody who builds netcfgd from a distribution toolchain,
//! which is a decision about who can compile this rather than a way of writing
//! four lines. Worth revisiting when the floor moves for a reason of its own.
//!
//! **Why `flock` and not a lock file created with `O_EXCL`.** The lock has to
//! survive whatever the holder does, including being killed: an `O_EXCL` file
//! outlives the process that made it, so every user of one needs a staleness
//! rule, and every staleness rule is a guess about how long the work takes. A
//! `flock` is held by the open file description and the kernel drops it when
//! the last descriptor closes, which a dying process does for free.

use std::fs::OpenOptions;
use std::io;
use std::os::fd::AsRawFd;
use std::path::Path;

/// An exclusive lock on a file, released when this is dropped.
///
/// The file's *contents* are not the point and are never read or written; what
/// is being locked is the name, so that two processes agree on who is midway
/// through a read-modify-write of something else.
#[derive(Debug)]
pub struct FileLock {
	/// Held because closing the descriptor is what releases the lock. Never
	/// read from or written to.
	file: std::fs::File,
}

impl FileLock {
	/// Take the exclusive lock, creating the file if it is not there.
	///
	/// **Blocks** until the lock is available, which is the behaviour that makes
	/// this useful: the caller wants the update to happen, not to be told that
	/// somebody else is updating. The critical sections it guards are a read, a
	/// serialisation and a rename.
	///
	/// # Errors
	///
	/// Returns an `io::Error` if the directory cannot be created, the file
	/// cannot be opened, or the lock cannot be taken. A caller that treats a
	/// failure as "carry on without the lock" is choosing the old behaviour
	/// deliberately and should say so.
	pub fn exclusive(path: &Path) -> io::Result<Self> {
		if let Some(parent) = path.parent() {
			std::fs::create_dir_all(parent)?;
		}
		// Opened for writing because a lock is a claim to change something, and
		// a descriptor opened read-only can still take `LOCK_EX` -- which would
		// make the mode say the opposite of what the caller means.
		let file = OpenOptions::new()
			.write(true)
			.create(true)
			.truncate(false)
			.open(path)?;

		// SAFETY: `flock` takes a descriptor and an integer and returns an
		// integer. No pointers. The descriptor is owned by `file`, which is
		// live for the rest of this function and is moved into the value
		// returned, so it cannot be closed while the lock is meant to be held.
		let rc = unsafe { libc::flock(file.as_raw_fd(), libc::LOCK_EX) };
		if rc < 0 {
			return Err(io::Error::last_os_error());
		}
		Ok(Self { file })
	}
}

impl Drop for FileLock {
	fn drop(&mut self) {
		// `LOCK_UN` rather than leaving it to the close that follows
		// immediately after. Both release it, and the explicit call is what
		// makes the release visible at the point the guard ends -- a reader
		// should not have to know that `File`'s own `Drop` is what unlocks
		// this. A failure here is not actionable and is not reported: the
		// close on the next line releases the lock regardless.
		// SAFETY: as in `exclusive`. The descriptor is still owned by
		// `self.file`, which is dropped after this.
		unsafe {
			libc::flock(self.file.as_raw_fd(), libc::LOCK_UN);
		}
	}
}

#[cfg(test)]
mod tests {
	use super::*;

	/// The lock is mutual between two descriptors of one file.
	///
	/// This is the property the whole module rests on and it is not the one
	/// POSIX record locks have: `fcntl` locks are owned by the *process*, so a
	/// second thread taking one would be handed it immediately and the guard
	/// would guard nothing. `flock` is owned by the open file description, so
	/// two `open` calls conflict whether or not they are in the same process --
	/// which is what makes a thread test meaningful evidence about two daemons.
	#[test]
	fn a_second_descriptor_waits_for_the_first() {
		let dir = netcfgd_testdir::TestDir::new("sys-lock");
		let path = dir.join("owned.lock");

		let held = FileLock::exclusive(&path).expect("the first lock");

		// Non-blocking, because the point is that it is *not* available. The
		// blocking call is what the daemon makes and is not what a test can
		// assert on without a deadline.
		let second = OpenOptions::new()
			.write(true)
			.create(true)
			.truncate(false)
			.open(&path)
			.expect("opened");
		// SAFETY: as in `exclusive`.
		let rc = unsafe { libc::flock(second.as_raw_fd(), libc::LOCK_EX | libc::LOCK_NB) };
		assert_eq!(rc, -1, "a second exclusive lock must not be granted");
		assert_eq!(
			io::Error::last_os_error().raw_os_error(),
			Some(libc::EWOULDBLOCK),
			"refused for the wrong reason"
		);

		drop(held);

		// SAFETY: as above.
		let rc = unsafe { libc::flock(second.as_raw_fd(), libc::LOCK_EX | libc::LOCK_NB) };
		assert_eq!(rc, 0, "the lock must be free once the guard is dropped");
	}
}

File: c/src/sys/lock.c
/*
 * lock.c -- `flock`, a deadline, and the directory the lock file sits in.
 *
 * The whole of the argument for `flock` over the alternatives is in lock.h.
 * What is here is the three things it needs: the parent directory, the open,
 * and the wait.
 */
#include "ncfg/lock.h"

#include <errno.h>
#include <fcntl.h>
#include <string.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <time.h>
#include <unistd.h>

/* How long to wait between attempts while a deadline is running. Short enough
 * that a lock released early is taken promptly, long enough that waiting costs
 * nothing measurable. */
#define LOCK_RETRY_MS 20L

/* Room for the parent of a lock path. */
#define LOCK_PATH_MAX 4096u

static long now_ms(void)
{
	struct timespec at;

	/* Monotonic, because a deadline measured against the wall clock is a
	 * deadline that moves when something sets the time -- which on a machine
	 * netcfgd is configuring is precisely what happens when the network
	 * comes up and NTP steps the clock. */
	if (clock_gettime(CLOCK_MONOTONIC, &at) != 0) {
		return 0;
	}
	return (long)at.tv_sec * 1000L + at.tv_nsec / 1000000L;
}

static void wait_ms(long milliseconds)
{
	struct timespec wanted;

	wanted.tv_sec = milliseconds / 1000L;
	wanted.tv_nsec = (milliseconds % 1000L) * 1000000L;
	(void)nanosleep(&wanted, NULL);
}

/*
 * Make the directories a lock path needs, as `mkdir -p` would.
 *
 * Each component is created and an existing one is not an error. Nothing here
 * removes anything: a directory that is in the way and is not a directory is
 * reported by the open that follows, which names the path.
 */
static int make_parents(const char *path, char *err, size_t err_size)
{
	char   work[LOCK_PATH_MAX];
	size_t length = strlen(path);
	size_t at;

	if (length >= sizeof(work)) {
		ncfg_error_set(err, err_size, "%s is too long a path for a lock", path);
		return 0;
	}
	memcpy(work, path, length + 1u);
	/* From 1, so a leading `/` is never treated as a component to create. */
	for (at = 1; at < length; at++) {
		if (work[at] != '/') {
			continue;
		}
		work[at] = '\0';
		if (mkdir(work, 0755) != 0 && errno != EEXIST) {
			ncfg_error_set(err, err_size, "cannot make %s: %s", work,
			    strerror(errno));
			return 0;
		}
		work[at] = '/';
	}
	return 1;
}

void ncfg_lock_init(ncfg_lock_t *lock)
{
	if (!lock) {
		return;
	}
	lock->fd = -1;
}

/* Open the file the lock is taken on. */
static int lock_open(const char *path, char *err, size_t err_size)
{
	int opened;

	if (!make_parents(path, err, err_size)) {
		return -1;
	}
	/* Write, create, and **not** truncate: the contents are not the point
	 * and another holder's descriptor is watching this same file. Opened
	 * for writing because a lock is a claim to change something. */
	opened = open(path, O_WRONLY | O_CREAT | O_CLOEXEC, 0644);
	if (opened < 0) {
		ncfg_error_set(err, err_size, "cannot open %s: %s", path, strerror(errno));
		return -1;
	}
	return opened;
}

int ncfg_lock_take(ncfg_lock_t *lock, const char *path, char *err, size_t err_size)
{
	int opened;

	if (!lock || !path) {
		ncfg_error_set(err, err_size, "taking a lock needs a path");
		return 0;
	}
	ncfg_lock_init(lock);
	opened = lock_open(path, err, err_size);
	if (opened < 0) {
		return 0;
	}
	while (flock(opened, LOCK_EX) != 0) {
		/* A signal arriving while blocked is not a lock somebody else
		 * holds. Asking again is what the caller wanted; reporting it
		 * would turn a `SIGCHLD` -- which netcfgd generates itself,
		 * every time it spawns a hook -- into a failed apply. */
		if (errno == EINTR) {
			continue;
		}
		ncfg_error_set(err, err_size, "cannot lock %s: %s", path, strerror(errno));
		(void)close(opened);
		return 0;
	}
	lock->fd = opened;
	return 1;
}

int ncfg_lock_take_within(ncfg_lock_t *lock, const char *path, long patience_ms, int *held,
    char *err, size_t err_size)
{
	long deadline;
	int  opened;

	if (held) {
		*held = 0;
	}
	if (!lock || !path) {
		ncfg_error_set(err, err_size, "taking a lock needs a path");
		return 0;
	}
	ncfg_lock_init(lock);
	opened = lock_open(path, err, err_size);
	if (opened < 0) {
		return 0;
	}
	deadline = now_ms() + (patience_ms > 0 ? patience_ms : 0);
	for (;;) {
		if (flock(opened, LOCK_EX | LOCK_NB) == 0) {
			lock->fd = opened;
			return 1;
		}
		if (errno != EWOULDBLOCK && errno != EINTR) {
			ncfg_error_set(err, err_size, "cannot lock %s: %s", path,
			    strerror(errno));
			(void)close(opened);
			return 0;
		}
		if (now_ms() >= deadline) {
			if (held) {
				*held = 1;
			}
			if (patience_ms >= 1000L) {
				ncfg_error_set(err, err_size,
				    "%s is held by another apply, which has not finished in %ld seconds",
				    path, patience_ms / 1000L);
			} else {
				ncfg_error_set(err, err_size,
				    "%s is held by another apply, which has not finished in %ld milliseconds",
				    path, patience_ms);
			}
			(void)close(opened);
			return 0;
		}
		wait_ms(LOCK_RETRY_MS);
	}
}

void ncfg_lock_release(ncfg_lock_t *lock)
{
	if (!lock || lock->fd < 0) {
		return;
	}
	/* `LOCK_UN` rather than leaving it to the close on the next line. Both
	 * release it, and a failure here is not actionable and is not reported:
	 * the close releases the lock regardless. */
	(void)flock(lock->fd, LOCK_UN);
	(void)close(lock->fd);
	lock->fd = -1;
}
